sqlmesh/core/model/definition.py

```diff
@@ -1003,6 +1003,18 @@ def columns_to_types(self) -> t.Optional[t.Dict[str, exp.DataType]]:
                 for select in query.selects
             }
 
+            if self.time_column:
+                time_column_name = self.time_column.column.output_name
+                if self._columns_to_types[time_column_name] == exp.DataType.build(
+                    exp.DataType.Type.UNKNOWN
+                ):
+                    # if the user has specified a time column, but we cant detect its type - assume its a timestamp
+                    # if we assume its "unknown" then in practice gets treated like a string and the incremental model delete queries
+                    # will happily try to filter a timestamp column based on a string literal, which is invalid in some engines
+                    self._columns_to_types[time_column_name] = exp.DataType.build(
+                        exp.DataType.Type.TIMESTAMP
```

Comment:
i think we need to think about the implications of this and test every single engine. the original design defaults to date string because those are common in hive / spark.

additionally this may not be the place to do this. we could probably just add this into the section where we add the time filter

Reply:
Agreed, although the existing test suite does pass no problem.

Even if we change where this is done, I guess the core question still remains - is it valid to default the column specifically designated as the incremental model ` time_column`  to `timestamp` if its type is unknown? I personally think it is; how could you have an incremental model based on a `time_column` that isnt a date or timestamp type?

Another approach might be to make `columns (.. )` on the model definition override the auto-detected column types rather than replace them entirely (which would also be a potentially breaking change I think).

Then I could just override the type of the `time_column` to be what I need and leave everything else alone

Reply:
well it’s valid because your time column could be a string, which is common in spark. and comparing a date string to a timestamp works fine in spark. it just doesn’t work in trino which has stricter types

Reply:
Is it detected as a string though? This would only change it to a timestamp if it was detected as "unknown". If it was detected as "varchar" or whatever the Spark string type is, it would remain as a string

Reply:
Based on this conversation, it feels like there's no one-size-fits-all solution, and the behavior largely depends on a specific engine. As such I'd suggest solving this for individual engines as well. We should probably introduce this in specific engine adapter implementations.

Reply:
Ok, no problem, I can make this specific to the Trino adapter instead.

There is another potential approach though - and that is allowing the `columns` property on the Model to be merged into `columns_to_types` without replacing them entirely.

Then I could just do:
```
MODEL (
  ...
  columns (
    ds timestamp
  )
);
```

What do you think of this approach?

This would also help some of my other models where part of the resultset is determined by a Trino table function which means that some of the types cannot be determined just by analysing the SQL. Right now I have to painstakingly specify every single column type, even the ones that can be inferred, but it would be nice not to have to :)